### PR TITLE
CLDR-17014 Handle VARIANT paths like LANGUAGE, SCRIPT, ...

### PR DESCRIPTION
**Pull request by btangmu**
-Use ExtraPaths instead of XMLSource.ResolvingSource.constructedItems

-Paths constructed in this way no longer have fallback values

-Revise test data localeDisplayName.txt to expect fonipa instead of FONIPA

-Remove three paths from common/main/no.xml that had INHERITANCE_MARKER for paths that had no value to inherit

CLDR-17014

- [ ] This PR completes the ticket.

ALLOW_MANY_COMMITS=true

### CODE REVIEW

File: tools/cldr-code/src/test/java/org/unicode/cldr/unittest/TestCLDRFile.java [hunk 1]
```diff
@@ -195,6 +195,7 @@ public void testExtraPaths() {
                                     || path.startsWith("//ldml/localeDisplayNames/scripts/script")
                                     || path.startsWith(
                                             "//ldml/localeDisplayNames/territories/territory")
+                                    || path.startsWith("//ldml/localeDisplayNames/variants/variant")
```
Comment (btangmu) [inline tools/cldr-code/src/test/java/org/unicode/cldr/unittest/TestCLDRFile.java:198]: This is in testExtraPaths. Adding the VARIANT paths here enables the test to pass, given that `german.getExtraPaths()` now includes VARIANT paths. PlaceholderStatus appears to depend on the file Placeholders.txt, but I don't understand what it means. It seems plausible that whatever PlaceholderStatus means, isn't changed by the fact that VARIANT paths are now implemented using the ExtraPaths mechanism and they no longer have code-fallback values.

File: common/testData/localeIdentifiers/localeDisplayName.txt [hunk 2]
```diff
@@ -356,7 +356,7 @@ es-Cyrl-MX; Spaans (Cyrillies, Meksiko)
 hi-Latn; Hindi (Latyn)
 nl-BE; Nederlands (België)
 nl-Latn-BE; Nederlands (Latyn, België)
-zh-Hans-fonipa; Chinees (Vereenvoudig, FONIPA)
+zh-Hans-fonipa; Chinees (Vereenvoudig, fonipa)
```
Comment (btangmu) [inline common/testData/localeIdentifiers/localeDisplayName.txt:359]: This change enables TestLocaleDisplay to pass. The all-caps version FONIPA presumably came from the code-fallback value for a path like `//ldml/localeDisplayNames/variants/variant[@type="FONIPA"]`. Since this PR removes that code-fallback, the value needs to come from elsewhere; it's not clear to me where it comes from. 
Maybe from the string "zh-Hans-fonipa" itself? This line in `TestLocaleDisplay` seems to be where it happens:

```
    String name = cldrFile.nameGetter().getNameFromParserOpt(ltp, nameOpt);
```

See `testInputLocales` in GenerateLocaleIDTestData.java. The value `Chinees (Vereenvoudig, fonipa)` seems no worse than `Chinees (Vereenvoudig, FONIPA)`.

File: tools/cldr-code/src/main/java/org/unicode/cldr/util/XMLSource.java [hunk 3]
```diff
@@ -1598,11 +1598,7 @@ public String getLocaleID() {
             Map<String, Set<String>> countries_zoneSet = sc.getCountryToZoneSet();
             Map<String, String> zone_countries = sc.getZoneToCountry();
             List<NameType> nameTypeList =
-                    List.of(
-                            NameType.VARIANT,
-                            NameType.CURRENCY,
-                            NameType.CURRENCY_SYMBOL,
-                            NameType.TZ_EXEMPLAR);
+                    List.of(NameType.CURRENCY, NameType.CURRENCY_SYMBOL, NameType.TZ_EXEMPLAR);
```
Comment (btangmu) [inline tools/cldr-code/src/main/java/org/unicode/cldr/util/XMLSource.java:1601]: Only 3 left

**Comment by btangmu** [comment]
Locally, all tests passed, but in this PR we have failures:

```
    TestPathHeadersAndValues
Error:  (TestPaths.java:199)  Error: Locale: nn,	 Value=null, 	Path: //ldml/localeDisplayNames/variants/variant[@type="RUMGR"],	 IsExtraPath: false

Error:  (TestPaths.java:199)  Error: Locale: nn,	 Value=null, 	Path: //ldml/localeDisplayNames/variants/variant[@type="LTG1929"],	 IsExtraPath: false

Error:  (TestPaths.java:199)  Error: Locale: nn,	 Value=null, 	Path: //ldml/localeDisplayNames/variants/variant[@type="LUNA1918"],	 IsExtraPath: false
```

That's a puzzle... This is the test:

```
    public void TestPathHeadersAndValues() {
        /*
         * Use the pathsSeen hash to keep track of which paths have
         * already been seen. Since the test checkPrettyPaths isn't really
         * locale-dependent, run it only once for each path, for the first
         * locale in which the path occurs.
         */
        Set<String> pathsSeen = new HashSet<>();
        CLDRFile englishFile = testInfo.getCldrFactory().make("en", true);
        PathHeader.Factory phf = PathHeader.getFactory(englishFile);
        Status status = new Status();
        final String exemptLocale = "sv";
        final String exemptPathIfLocale =
                "//ldml/dates/calendars/calendar[@type=\"dangi\"]/dateTimeFormats/availableFormats/dateFormatItem[@id=\"yMd\"]";

        for (String locale : getLocalesToTest()) {
            boolean isExemptLocale = locale.equals(exemptLocale);

            if (!StandardCodes.isLocaleAtLeastBasic(locale)) {
                continue;
            }
            CLDRFile file = testInfo.getCLDRFile(locale, true);
            logln("Testing path headers and values for locale => " + locale);
            final Collection<String> extraPaths = file.getExtraPaths();

            for (Iterator<String> it = file.iterator(); it.hasNext(); ) {
                String path = it.next();
                if (isExemptLocale && path.equals(exemptPathIfLocale)) {
                    logKnownIssue("CLDR-17849", "Can't reproduce locally");
                    continue;
                }
                if (extraPaths.contains(path)) {
                    checkFullpathValue(path, file, locale, status, true /* extra path */);
                } else {
                    checkFullpathValue(path, file, locale, status, false /* not extra path */);
                }
                if (!pathsSeen.contains(path)) {
                    pathsSeen.add(path);
                    checkPrettyPaths(path, phf);
                }
            }
        }
    }
```

File: tools/cldr-code/src/test/java/org/unicode/cldr/unittest/TestCLDRFile.java (hunk 1 above)
Comment (macchiati) [reply tools/cldr-code/src/test/java/org/unicode/cldr/unittest/TestCLDRFile.java:198]: PlaceholderStatus tells whether a value for that path can have placeholders or not. (Cf. https://cldr.unicode.org/development/updating-dtds#placeholders)
```   
public enum PlaceholderStatus {
        DISALLOWED("No placeholders allowed."), //
        REQUIRED("Specific number of placeholders allowed."), //
        LOCALE_DEPENDENT("Some placeholders may be omitted in certain locales"), //
        MULTIPLE(
                "May have multiple instances of the same placeholder, eg “{0} cats and {0} dogs”."), //
        OPTIONAL(
                "Any specific placeholder is optional (and non-numeric); there must be at least one.") //
    ;
```
It is data-driven, from Placeholders.txt. If the path doesn't match in the file (which it doesn't for variants), the value is DISALLOWED.

**Comment by btangmu** [comment]
The second commit might allow `TestPathHeadersAndValues` to pass. However, that wasn't necessary locally, so I'm mystified. There's a dubious smell about the exceptional handling for "extra" paths. See https://unicode-org.atlassian.net/browse/CLDR-18217

File: tools/cldr-code/src/test/java/org/unicode/cldr/unittest/TestCLDRFile.java (hunk 1 above)
Comment (btangmu) [reply tools/cldr-code/src/test/java/org/unicode/cldr/unittest/TestCLDRFile.java:198]: Thanks! That makes sense. So the change to `testExtraPaths` seems appropriate. As that list of `path.startsWith(...)` conditions grows, the test should maybe be refactored to move that part into a subroutine.

**Comment by btangmu** [comment]
Disappointingly `TestPathHeadersAndValues` still fails after the 2nd commit

**Comment by macchiati** [comment]
The main goal of that test is at line 159

```
     * For the given path and CLDRFile, check that fullPath, value, and source are all non-null.
     *
     * <p>Allow null value for some exceptional extra paths.
```
The way it should work is that you would only get null values for paths that were in extraPaths. And that should still be true.

So looking a the error line, the question is, how did that path show up in the cldr file nn with a null value if it wasn't in extraPaths??

```
(TestPaths.java:199)  Error: Locale: nn,	 Value=null, 	Path: //ldml/localeDisplayNames/variants/variant[@type="RUMGR"],	 IsExtraPath: false
```

**Comment by btangmu** [comment]
Yes, and another question is, why does the test pass locally on my macBook but not in this PR?

Is there anything unusual about the locale nn?

**Comment by btangmu** [comment]
Interesting that nn inherits from no. Here is the RUMGR path in ST, in nn, no, and root:

https://st.unicode.org/cldr-apps/v#/nn/Locale_Variants/44dcd5a74b5a5f79
https://st.unicode.org/cldr-apps/v#/no/Locale_Variants/44dcd5a74b5a5f79
https://st.unicode.org/cldr-apps/v#/root/Locale_Variants/44dcd5a74b5a5f79

**Comment by btangmu** [comment]
```
$ egrep RUMGR common/main/*.xml
common/main/bal_Latn.xml:			<variant type="RUMGR">RÓMGR</variant>
common/main/bew.xml:			<variant type="RUMGR" draft="unconfirmed">Graubenderlan</variant>
common/main/br.xml:			<variant type="RUMGR">romañcheg Grischun</variant>
common/main/ca.xml:			<variant type="RUMGR" draft="contributed">interomanx</variant>
common/main/cv.xml:			<variant type="RUMGR" draft="unconfirmed">RUMGR локализаци</variant>
common/main/eu.xml:			<variant type="RUMGR">Rumgr</variant>
common/main/fi.xml:			<variant type="RUMGR">retoromaanin rumantsch grischun -muoto</variant>
common/main/fo.xml:			<variant type="RUMGR">rumgr</variant>
common/main/fr.xml:			<variant type="RUMGR" draft="unconfirmed">standard des Grisons</variant>
common/main/fy.xml:			<variant type="RUMGR" draft="contributed">Rumgr</variant>
common/main/ga.xml:			<variant type="RUMGR">Rumgr</variant>
common/main/gd.xml:			<variant type="RUMGR">Rumantsch Grischun</variant>
common/main/he.xml:			<variant type="RUMGR" draft="contributed">רמגר</variant>
common/main/hu.xml:			<variant type="RUMGR" draft="contributed">Rumgr</variant>
common/main/ka.xml:			<variant type="RUMGR" draft="unconfirmed">გრიშუნული რეტორომანული</variant>
common/main/kk.xml:			<variant type="RUMGR" draft="contributed">РУМГР</variant>
common/main/kk_Arab.xml:			<variant type="RUMGR" draft="unconfirmed">رۋمگر</variant>
common/main/lo.xml:			<variant type="RUMGR">ຣຳກຣ໌</variant>
common/main/ms.xml:			<variant type="RUMGR">RUMANTSCH GRISCHUN</variant>
common/main/nl.xml:			<variant type="RUMGR" draft="contributed">Rumgr</variant>
common/main/no.xml:			<variant type="RUMGR">↑↑↑</variant>
common/main/ro.xml:			<variant type="RUMGR" draft="contributed">Rumgr</variant>
common/main/sc.xml:			<variant type="RUMGR">istandard de sos Grisones</variant>
common/main/sv.xml:			<variant type="RUMGR" draft="contributed">grischun-dialekt</variant>
common/main/th.xml:			<variant type="RUMGR" draft="contributed">รัมกร์</variant>
common/main/to.xml:			<variant type="RUMGR" draft="unconfirmed">lea fakafeituʻu fakakilisone</variant>
common/main/yue.xml:			<variant type="RUMGR" draft="contributed">羅曼什文</variant>
common/main/yue_Hans.xml:			<variant type="RUMGR" draft="contributed">罗曼什文</variant>
common/main/zh.xml:			<variant type="RUMGR">罗曼什文</variant>
common/main/zh_Hant.xml:			<variant type="RUMGR" draft="contributed">羅曼什文</variant>
common/main/zh_Hant_HK.xml:			<variant type="RUMGR">↑↑↑</variant>
common/main/zu.xml:			<variant type="RUMGR">i-RUMGR</variant>
```

no.xml has ↑↑↑, while it's in neither nn.xml nor root.xml

**Comment by btangmu** [comment]
That seems to have fixed the RUMGR failure! And the same should fix the other 2:

```
    TestPathHeadersAndValues
Error:  (TestPaths.java:199)  Error: Locale: nn,	 Value=null, 	Path: //ldml/localeDisplayNames/variants/variant[@type="LTG1929"],	 IsExtraPath: false
Error:  (TestPaths.java:199)  Error: Locale: nn,	 Value=null, 	Path: //ldml/localeDisplayNames/variants/variant[@type="LUNA1918"],	 IsExtraPath: false
 (30.239s) FAILED (2 failure(s))
```

```
$ egrep LTG1929 common/main/*.xml
common/main/bal_Latn.xml:			<variant type="LTG1929">LTG-1929</variant>
common/main/cv.xml:			<variant type="LTG1929" draft="unconfirmed">LTG1929 локализаци</variant>
common/main/fo.xml:			<variant type="LTG1929">ltg1929</variant>
common/main/gd.xml:			<variant type="LTG1929">Litreachadh na Latgailise 1929</variant>
common/main/ie.xml:			<variant type="LTG1929" draft="unconfirmed">ortografie de 1929</variant>
common/main/kk_Arab.xml:			<variant type="LTG1929" draft="unconfirmed">لتگ1929</variant>
common/main/no.xml:			<variant type="LTG1929">↑↑↑</variant>
common/main/sc.xml:			<variant type="LTG1929">ortografia de sa limba latgaliana de su 1929</variant>
common/main/to.xml:			<variant type="LTG1929" draft="unconfirmed">lea fakalatakalīsu (motuʻa)</variant>

$ egrep LUNA1918 common/main/*.xml
common/main/bal_Latn.xml:			<variant type="LUNA1918">LUNÁ-1918</variant>
common/main/bew.xml:			<variant type="LUNA1918" draft="unconfirmed">Èjaan Rus Perobahan tahon 1918</variant>
common/main/br.xml:			<variant type="LUNA1918">reizhskrivadur rusianek goude 1917</variant>
common/main/bs.xml:			<variant type="LUNA1918">Luna1918</variant>
common/main/ca.xml:			<variant type="LUNA1918" draft="contributed">luna 1918</variant>
common/main/cv.xml:			<variant type="LUNA1918" draft="unconfirmed">LUNA1918 локализаци</variant>
common/main/eu.xml:			<variant type="LUNA1918">Luna1918</variant>
common/main/fi.xml:			<variant type="LUNA1918">venäjän ortografia 1918</variant>
common/main/fo.xml:			<variant type="LUNA1918">luna1918</variant>
common/main/fr.xml:			<variant type="LUNA1918" draft="unconfirmed">orthographe russe réformée de 1918</variant>
common/main/fy.xml:			<variant type="LUNA1918" draft="contributed">Luna1918</variant>
common/main/ga.xml:			<variant type="LUNA1918">Luna1918</variant>
common/main/gd.xml:			<variant type="LUNA1918">Litreachadh na Ruisise às dèidh 1917</variant>
common/main/he.xml:			<variant type="LUNA1918" draft="contributed">לונה1918</variant>
common/main/hu.xml:			<variant type="LUNA1918" draft="contributed">Luna1918</variant>
common/main/ie.xml:			<variant type="LUNA1918" draft="unconfirmed">modern ortografie</variant>
common/main/ka.xml:			<variant type="LUNA1918" draft="unconfirmed">1918 წლის რუსილი</variant>
common/main/kk.xml:			<variant type="LUNA1918" draft="contributed">ЛУНА1918</variant>
common/main/kk_Arab.xml:			<variant type="LUNA1918" draft="unconfirmed">لۋنا1998</variant>
common/main/ko.xml:			<variant type="LUNA1918" draft="contributed">루나1918</variant>
common/main/lo.xml:			<variant type="LUNA1918">ລູນ່າ1918</variant>
common/main/ms.xml:			<variant type="LUNA1918">ORTOGRAFI RUSIA SELEPAS 1917</variant>
common/main/nl.xml:			<variant type="LUNA1918" draft="contributed">Russische spelling van 1917</variant>
common/main/no.xml:			<variant type="LUNA1918">↑↑↑</variant>
common/main/ro.xml:			<variant type="LUNA1918" draft="contributed">Luna1918</variant>
common/main/sc.xml:			<variant type="LUNA1918">ortografia russa riformada de su 1918</variant>
common/main/so.xml:			<variant type="LUNA1918">LUUNA1918</variant>
common/main/sv.xml:			<variant type="LUNA1918">1918 års stavning</variant>
common/main/th.xml:			<variant type="LUNA1918" draft="contributed">ลูน่า 1918</variant>
common/main/to.xml:			<variant type="LUNA1918" draft="unconfirmed">lea fakalūsia-soviete</variant>
common/main/yue.xml:			<variant type="LUNA1918" draft="contributed">俄羅斯文拼字（1917年後）</variant>
common/main/yue_Hans.xml:			<variant type="LUNA1918" draft="contributed">俄罗斯文拼字（1917年后）</variant>
common/main/zh.xml:			<variant type="LUNA1918">俄文拼字（1918年起）</variant>
common/main/zh_Hant.xml:			<variant type="LUNA1918" draft="contributed">俄羅斯文拼字（1917年後）</variant>
common/main/zh_Hant_HK.xml:			<variant type="LUNA1918">↑↑↑</variant>
common/main/zu.xml:			<variant type="LUNA1918">i-LUNA1918</variant>
```

**Review by macchiati — COMMENTED** [review]
Great work!

Sounds like we also need a test: whenever a locale path has three up arrows, then it's Bailey value must not be null. Might be easy to add to the test that's been failing.

**Review by macchiati — CHANGES_REQUESTED** [review]
I want to make sure that you generated the change in the test data file  by running the test generation tool, **not** with a spot fix that will be overwritten the next time the tool is run. Otherwise, the problem could recur when the test generation tool is run as part of the BRS.

**Comment by btangmu** [comment]
> I want to make sure that you generated the change in the test data file by running the test generation tool, **not** with a spot fix that will be overwritten the next time the tool is run. Otherwise, the problem could recur when the test generation tool is run as part of the BRS.

It's a good thing you noticed that! I didn't know the file was generated. I'll figure out what tool generates it, and update this PR with that change...

**Comment by btangmu** [comment]
@macchiati I ran
```
mvn -DCLDR_DIR=$(pwd) --file=tools/pom.xml -pl cldr-code compile -DskipTests=true exec:java -Dexec.mainClass=org.unicode.cldr.tool.GenerateLocaleIDTestData 
```
and the file localeDisplayName.txt did get regenerated, with lowercase for "fonipa". Evidently the changes already in this PR are sufficient to make GenerateLocaleIDTestData do the right thing next time it's run.

**Comment by btangmu** [comment]
> Sounds like we also need a test: whenever a locale path has three up arrows, then it's Bailey value must not be null. Might be easy to add to the test that's been failing.

I just made a new ticket for that: https://unicode-org.atlassian.net/browse/CLDR-18294

**Review by macchiati — APPROVED** [review]
